src/python/pants/build_graph/address.py

```diff
@@ -244,18 +244,20 @@ class BuildFileAddress(Address):
   :API: public
   """
 
-  def __init__(self, build_file, target_name=None):
+  def __init__(self, build_file=None, rel_path=None, target_name=None):
     """
     :param build_file: The build file that contains the object this address points to.
     :type build_file: :class:`pants.base.build_file.BuildFile`
+    :param string rel_path: The path relative to root_dir where the BUILD file is located.
     :param string target_name: The name of the target within the BUILD file; defaults to the default
                                target, aka the name of the BUILD file parent dir.
 
     :API: public
     """
-    spec_path = os.path.dirname(build_file.relpath)
+    spec_path = os.path.dirname(rel_path if rel_path else build_file.relpath)
     super(BuildFileAddress, self).__init__(spec_path=spec_path,
                                            target_name=target_name or os.path.basename(spec_path))
+    self.rel_path = rel_path if rel_path else build_file.relpath
```

Comment:
seems like you could tighten up the duplication here and use boolean logic vs a ternary expression to simplify:

e.g.

```
rel_path = rel_path or build_file.relpath
spec_path = os.path.dirname(rel_path)
...
self.rel_path = rel_path
...
```